Repo Gardening: clean up issue contents before to send to OpenAI

jeherve opened the pull request:
## Proposed changes:

When analysing the contents of an issue to find relevant labels, we want to do 2 things:

- Remove any links, as they do not provide relevant information to OpenAI.
- If the issue has very little text (less than 100 characters, i.e. only one sentence), we cannot really figure out labels based on that alone.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* p1731934295971299-slack-C048CUFRGFQ

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

This can only be tested in a fork. Here are 2 issues, one with content that was analysed, one that wasn't:

- https://github.com/jeherve/jetpack/issues/175
- https://github.com/jeherve/jetpack/issues/176

jeherve committed:
Repo Gardening: clean up issue contents before to send to OpenAI

When analysing the contents of an issue to find relevant labels, we want to do 2 things:

- Remove any links, as they do not provide relevant information to OpenAI.
- If the issue has very little text (less than 100 characters, i.e. only one sentence), we cannot really figure out labels based on that alone.

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/ai-labeling.js:
```diff
@@ -94,6 +94,27 @@ Example response format:
+ */
+function cleanIssueContent( content ) {
+	// Remove links in the format [link text](url)
+	content = content.replace( /\[.*?\]\(.*?\)/g, '' );
```
Probably you'd want to keep the link text, just not the url?

jeherve replied:
That's a good point, I didn't think of that. What do you think of 788d7abc98e07998a766501dc24ca783cd027bca?

jeherve committed:
Keep links' text

See https://github.com/Automattic/jetpack/pull/40269/files#r1850564592

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/ai-labeling.js:
```diff
@@ -94,6 +94,27 @@ Example response format:
+ */
+function cleanIssueContent( content ) {
+	// Remove links in the format [link text](url), but keep the link text.
+	content = content.replace( /\[(.*?)\]\(https?:\/\/\S+\)/g, '$1' );
```
Edge case: if the content has something like `text [xyz] text [foo](https://example.com) text`, it'll unexpectedly get turned into `text xyz] text [foo text`.

Another one: `text [foo](https://example.com)-[bar](https://example.com) text` (i.e. no space between the links) gets unexpectedly turned into `text foo text` rather than `text foo-bar text`.

jeherve replied:
Ah yes, I need to be more exact. What do you think of bcf074a00ed7dc4aa55803cf860a5b6217714b24?

jeherve committed:
Catch more content in markdown formatted differently

See https://github.com/Automattic/jetpack/pull/40269/files#r1850614860

Example input:

```
text [xyz] text [foo](https://example.com) text

text [foo](https://example.com)-[bar](https://example.com) text

text <a href="https://example.com">foo</a> text

text <a href="https://example.com">foo</a>-<a href="https://example.com">bar</a> text
```

Output will be:

```
text xyz text foo text

text foo-bar text

text foo text

text foo-bar text
```

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/ai-labeling-conditions-clean-content

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/ai-labeling.js:
```diff
@@ -94,6 +94,27 @@ Example response format:
+ */
+function cleanIssueContent( content ) {
+	// Remove links in the format [link text](url), but keep the link text.
+	content = content.replace( /\[(.*?)\](?:\([^)]+\))?/g, '$1' );
```
If someone does a link like

> [Link text][1] and [another link][example.org]
>
> [1]: https://example.com
> [example.org]: https://example.org

```md
[Link text][1] and [another link][example.org]

[1]: https://example.com
[example.org]: https://example.org
```
it'll turn into
```
Link text1 and another linkexample.org

1: this link with more information
example.org: this link with more information
```

Also worth noting is that without the corresponding "footnote" it doesn't get linkified at all:

> [just brackets]

> [broken link][2]

Then there's the possibility of stuff like this

<details><summary>

> [huh][what](https://example.com)

</summary>

```md
[huh][what](https://example.com)
```

</details>

and

<details><summary>

> [huh][what][ok] [huh][ok][what]
>
> [ok]: https://example.com

</summary>

```md
[huh][what][ok] [huh][ok][what]

[ok]: https://example.com
```

</details>

jeherve replied:
I'm thinking we may be okay leaving those edge cases out, since that's not format that will commonly used in our issues here at Automattic, or overall in GitHub issues imo.

anomiex reviewed (approved):
Other than increasingly convoluted edge cases with markdown link syntax, seems ok. 🤷
